tap_appstore/client.py

```diff
@@ -18,13 +23,138 @@
+        return date + timedelta(days=1)
+
+    def update_stream_state(self, date):
+        """Update the stream state with the new date."""
+        self.stream_state['start_date'] = date.strftime('%Y-%m-%d')
+        self.logger.info(f"Updating state, new start date is {self.stream_state['start_date']}")
+
+    def get_records(self, context: dict = None):
+        """Return a generator of record-type dictionary objects."""
+        start_date = self.get_start_date(default_date='2024-01-01')
+        if not start_date:
+            logger.error("Start date could not be determined.")
+            return
+
+        start_date = start_date.replace(tzinfo=None)
+        date_limit = datetime.now().replace(tzinfo=None) - timedelta(days=2)
+        line_id = 0
+
+        while start_date <= date_limit:
+
+            try:
+                logger.info(f"report_date: {self.get_report_date(start_date)}")
+                all_data = self.download_data(start_date, self.api)
+
+                data_io = StringIO(all_data)
+
+                # Rename column names from 'App Name' -> 'app_name'
+                first_line = data_io.readline().strip()
+                fieldnames = [col.strip().replace(' ', '_').lower() for col in first_line.split('\t')]
+
+                reader = csv.DictReader(data_io, delimiter='\t', fieldnames=fieldnames)
+
+                for record in reader:
+                    line_id += 1
+                    record['_line_id'] = line_id
+                    record['_time_extracted'] = datetime.utcnow().strftime('%Y-%m-%dT%H:%M:%SZ')
+                    record['_api_report_date'] = self.get_report_date(start_date)
+
+                    processed_record = self.process_record(record)
+                    if processed_record is not None:
+                        yield processed_record
+
+                    yield record
+
+            except APIError as e:
+                logger.error(f'Error during download report {self.name}.\n{e}')
+
+            start_date = self.increment_date(start_date)
+            self.update_stream_state(start_date)
+
+
+    @staticmethod
+    def convert_date(date_str, date_format='%Y-%m-%d'):
+        """Converts date string to ISO format based on the given date format, defaulting to '%Y-%m-%d'.
+           Returns None if the date string is invalid or empty.
+        """
+
+        if not date_str:
+            logger.warning(f"Empty or invalid date string provided; cannot convert using format {date_format}")
+            return None
+
+        date_str = date_str.strip()
+        if not date_str:
+            logger.warning(f"Empty or invalid date string provided; cannot convert using format {date_format}")
+            return None
+
+        try:
+            return datetime.strptime(date_str, date_format).isoformat()
+        except ValueError as e:
+            logger.error(f"Date conversion error for date: '{date_str}' with format '{date_format}' | Error: {str(e)}")
+            return None
```

Comment:
Let the singer sdk convert the date, if we need to adjust something like remove an empty string we can do it in the post-processing.

Reply:
I can't seem to get the Singer SDK to automatically convert the date. Will look into this more in our pairing session.

Reply:
@joaopamaral I tried removing the call to this `convert_date` function for `th.DateTime` fields, but I am seeing errors like this when I do that:

`2024-05-16T11:04:51.094754Z [info     ] ValueError: Could not parse value '03/03/2024' for field 'start_date'. cmd_type=loader name=target-parquet run_id=2b158109-3fa5-457f-a9f0-d0219ca18ab0 state_id=mahangu_test_29 stdio=stderr`